### PR TITLE
Fix Election when ForceNone V1

### PR DESCRIPTION
**Pull request by kianenigma**
This is one approach to fix the current polkadot situation. 

This is based on the assumption **that we will still call into `new_session()`. This is needed in order to clean things up**. See the tests as well. 

I might write another V2 solution as well in which we don't even open the election window if ForceNone is set. Although: 
1. I am not sure if this is logically correct. Perhaps staking should work as seamless as possible even when ForceNone. 
2. This requires a data migration. We need to clean some storage items. If someone can double check this solution, it is easier.

### CODE REVIEW

**Comment by kianenigma** [comment]
(wait up -- still wrong)

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -2564,21 +2564,24 @@ impl<T: Trait> Module<T> {
...
+		if era_length == T::SessionsPerEra::get() - 1 {
+			IsCurrentSessionFinal::put(true);
+		} else if era_length >= T::SessionsPerEra::get() {
+			// Should only happen when we are ready to trigger an era but we have ForceNone,
+			// otherwise previous arm would short circuit.
+			Self::close_election_window();
+		}
+		return None
+	},
+}
```
Comment (thiolliere) [inline frame/staking/src/lib.rs:2582]: indentation

**Comment by thiolliere** [comment]
> This requires a data migration. We need to clean some storage items. If someone can double check this solution, it is easier.

about migration, the next new_session change will clean it anyway with `close_election_window`, no ?

also, indeed as you said if ForceNone maybe we shouldn't do `IsCurrentSessionFinal::put(true);` at all. This could be slightly better as it is very probable that the session is actually not final.

**Comment by kianenigma** [comment]
Companion build fail should get fixed after https://github.com/paritytech/polkadot/pull/1160

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -2569,16 +2569,19 @@ impl<T: Trait> Module<T> {
 				Forcing::ForceAlways => (),
 				Forcing::NotForcing if era_length >= T::SessionsPerEra::get() => (),
 				_ => {
-					// not forcing, not a new era either. If final, set the flag.
-					if era_length + 1 >= T::SessionsPerEra::get() {
+					// Either `ForceNone`, or `NotForcing && era_length < T::SessionsPerEra::get()`.
+					if era_length == T::SessionsPerEra::get() - 1 {
```
Comment (gavofyork) [inline frame/staking/src/lib.rs:2573]:
```suggestion
					if era_length + 1 == T::SessionsPerEra::get() {
```

File: frame/staking/src/lib.rs [hunk 3]
```diff
@@ -2569,16 +2569,19 @@ impl<T: Trait> Module<T> {
 				Forcing::ForceAlways => (),
 				Forcing::NotForcing if era_length >= T::SessionsPerEra::get() => (),
 				_ => {
-					// not forcing, not a new era either. If final, set the flag.
-					if era_length + 1 >= T::SessionsPerEra::get() {
```
Comment (gavofyork) [inline frame/staking/src/lib.rs:2573]: why changing this at all?

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -2912,6 +2915,18 @@ impl<T: Trait> Module<T> {
 		maybe_new_validators
 	}
 
+
+	/// Remove all the storage items associated with the election. Should only once election round
+	/// is done.
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:2920]:
```suggestion
	/// Remove all the storage items associated with the election.
```

**Pull request merged**